File: qa-portal/frontend/lib/core/models/execution.dart
class TestExecutionSummary {
  final String id;
  final String name;
  final String? version;
  final String environment;
  final String? startedAt;
  final String? finishedAt;
  final int total;
  final int passed;
  final int failed;
  final int blocked;
  final int notApplicable;
  final int pending;

  const TestExecutionSummary({
    required this.id,
    required this.name,
    this.version,
    required this.environment,
    this.startedAt,
    this.finishedAt,
    required this.total,
    required this.passed,
    required this.failed,
    required this.blocked,
    required this.notApplicable,
    required this.pending,
  });

  factory TestExecutionSummary.fromJson(Map<String, dynamic> json) {
    return TestExecutionSummary(
      id: json['id'] as String,
      name: json['name'] as String,
      version: json['version'] as String?,
      environment: json['environment'] as String,
      startedAt: json['started_at'] as String?,
      finishedAt: json['finished_at'] as String?,
      total: json['total'] as int,
      passed: json['passed'] as int,
      failed: json['failed'] as int,
      blocked: json['blocked'] as int,
      notApplicable: json['not_applicable'] as int,
      pending: json['pending'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'version': version,
      'environment': environment,
      'started_at': startedAt,
      'finished_at': finishedAt,
      'total': total,
      'passed': passed,
      'failed': failed,
      'blocked': blocked,
      'not_applicable': notApplicable,
      'pending': pending,
    };
  }
}

class TestExecution {
  final String id;
  final String projectId;
  final String name;
  final String? version;
  final String environment;
  final String? startedAt;
  final String? finishedAt;
  final List<ExecutionResult> results;

  const TestExecution({
    required this.id,
    required this.projectId,
    required this.name,
    this.version,
    required this.environment,
    this.startedAt,
    this.finishedAt,
    required this.results,
  });

  factory TestExecution.fromJson(Map<String, dynamic> json) {
    return TestExecution(
      id: json['id'] as String,
      projectId: json['project_id'] as String,
      name: json['name'] as String,
      version: json['version'] as String?,
      environment: (json['environment'] as String?) ?? '',
      startedAt: json['started_at'] as String?,
      finishedAt: json['finished_at'] as String?,
      results: (json['results'] as List<dynamic>? ?? [])
          .map((r) => ExecutionResult.fromJson(r as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'project_id': projectId,
      'name': name,
      'version': version,
      'environment': environment,
      'started_at': startedAt,
      'finished_at': finishedAt,
      'results': results.map((r) => r.toJson()).toList(),
    };
  }
}

class ExecutionResult {
  final String id;
  final String executionId;
  final String testCaseId;
  final String? assignedTo;
  final String status;
  final String? notes;
  final String? route;
  final String? executedAt;
  final int? durationSeconds;
  final Map<String, dynamic>? testCase;
  final List<dynamic>? screenshots;

  const ExecutionResult({
    required this.id,
    required this.executionId,
    required this.testCaseId,
    this.assignedTo,
    required this.status,
    this.notes,
    this.route,
    this.executedAt,
    this.durationSeconds,
    this.testCase,
    this.screenshots,
  });

  factory ExecutionResult.fromJson(Map<String, dynamic> json) {
    return ExecutionResult(
      id: json['id'] as String,
      executionId: json['execution_id'] as String,
      testCaseId: json['test_case_id'] as String,
      assignedTo: json['assigned_to'] as String?,
      status: json['status'] as String,
      notes: json['notes'] as String?,
      route: json['route'] as String?,
      executedAt: json['executed_at'] as String?,
      durationSeconds: json['duration_seconds'] as int?,
      testCase: json['test_case'] as Map<String, dynamic>?,
      screenshots: json['screenshots'] as List<dynamic>?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'execution_id': executionId,
      'test_case_id': testCaseId,
      'assigned_to': assignedTo,
      'status': status,
      'notes': notes,
      'route': route,
      'executed_at': executedAt,
      'duration_seconds': durationSeconds,
      'test_case': testCase,
      'screenshots': screenshots,
    };
  }
}

File: qa-portal/backend/app/models/execution.py
from sqlalchemy import Column, String, Text, Integer, DateTime, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from datetime import datetime, timezone
import uuid
import enum

from app.database import Base


class ExecutionResultStatus(str, enum.Enum):
    pending = "pending"
    passed = "passed"
    failed = "failed"
    blocked = "blocked"
    not_applicable = "not_applicable"


class ExecutionModule(Base):
    __tablename__ = "execution_modules"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    execution_id = Column(UUID(as_uuid=True), ForeignKey("test_executions.id", ondelete="CASCADE"), nullable=False)
    module_id = Column(UUID(as_uuid=True), ForeignKey("modules.id", ondelete="CASCADE"), nullable=False)

    execution = relationship("TestExecution", back_populates="selected_modules")
    module = relationship("Module")


class TestExecution(Base):
    __tablename__ = "test_executions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    name = Column(String(200), nullable=False)
    version = Column(String(50), nullable=True)
    environment = Column(String(100), nullable=True)
    started_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    finished_at = Column(DateTime(timezone=True), nullable=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)

    project = relationship("Project", back_populates="executions")
    creator = relationship("User", foreign_keys=[created_by])
    results = relationship("TestExecutionResult", back_populates="execution", cascade="all, delete-orphan")
    selected_modules = relationship("ExecutionModule", back_populates="execution", cascade="all, delete-orphan")


class TestExecutionResult(Base):
    __tablename__ = "test_execution_results"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    execution_id = Column(UUID(as_uuid=True), ForeignKey("test_executions.id", ondelete="CASCADE"), nullable=False)
    test_case_id = Column(UUID(as_uuid=True), ForeignKey("test_cases.id"), nullable=False)
    assigned_to = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    status = Column(SAEnum(ExecutionResultStatus), nullable=False, default=ExecutionResultStatus.pending)
    notes = Column(Text, nullable=True)
    executed_at = Column(DateTime(timezone=True), nullable=True)
    executed_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    duration_seconds = Column(Integer, nullable=True)

    execution = relationship("TestExecution", back_populates="results")
    test_case = relationship("TestCase")
    assignee = relationship("User", foreign_keys=[assigned_to])
    executor = relationship("User", foreign_keys=[executed_by])
    screenshots = relationship("Screenshot", back_populates="execution_result", cascade="all, delete-orphan", order_by="Screenshot.order")
    bugs = relationship("Bug", back_populates="execution_result", cascade="all, delete-orphan")


class Screenshot(Base):
    __tablename__ = "screenshots"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    execution_result_id = Column(UUID(as_uuid=True), ForeignKey("test_execution_results.id", ondelete="CASCADE"), nullable=False)
    file_path = Column(String(512), nullable=True)
    file_name = Column(String(255), nullable=False)
    wasabi_file_id = Column(String(255), nullable=True)
    order = Column(Integer, nullable=False, default=0)
    taken_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    taken_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)

    execution_result = relationship("TestExecutionResult", back_populates="screenshots")
    taker = relationship("User", foreign_keys=[taken_by])
